domain/application/state/application.go

```diff
@@ -1511,57 +1513,24 @@ func (st *State) GetCharmIDByApplicationName(ctx context.Context, name string) (
 		return "", errors.Capture(err)
 	}
 
-	query, err := st.Prepare(`
-SELECT &applicationCharmUUID.*
-FROM application
-WHERE uuid = $applicationID.uuid
-	`, applicationCharmUUID{}, applicationID{})
-	if err != nil {
-		return "", errors.Errorf("preparing query for application %q: %w", name, err)
-	}
-
-	var result charmID
+	var result corecharm.ID
 	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		appUUID, err := st.lookupApplication(ctx, tx, name)
 		if err != nil {
 			return errors.Errorf("looking up application %q: %w", name, err)
 		}
 
-		appIdent := applicationID{ID: appUUID}
-
-		var charmIdent applicationCharmUUID
-		if err := tx.Query(ctx, query, appIdent).Get(&charmIdent); err != nil {
-			if errors.Is(err, sqlair.ErrNoRows) {
-				return errors.Errorf("application %s: %w", name, applicationerrors.ApplicationNotFound)
-			}
-			return errors.Errorf("getting charm for application %q: %w", name, err)
-		}
-
-		// If the charmUUID is empty, then something went wrong with adding an
-		// application.
-		if charmIdent.CharmUUID == "" {
-			// Do not return a CharmNotFound error here. The application is in
-			// a broken state. There isn't anything we can do to fix it here.
-			// This will require manual intervention.
-			return errors.Errorf("application is missing charm")
-		}
```

Comment:
This can never happen since we now have a not null foreign key from applications to charms